### PR TITLE
Add control features

### PR DESCRIPTION
**Pull request by ellwiii**
This is what I have so far, would appreciate any comments and directions:
* for Android Google Maps myLocation button - it does not work without granting location permissions. Requesting location permission seems to be a fairly complicated topic for me, so I am still doing research on it 🥲
* for iOS MapKit myLocation button I've attempted to enable it (also to pick better than default placing and style) but I'm pretty sure it could be done better
* for iOS MapKit levelPicker - there is not much information I've found on this topic, still not clear to me if it is possible to set this control button without having an indoor map implemented in the first place

### CODE REVIEW

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 1]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
       prop("mapType") { view: GoogleMapsView, mapType: MapType ->
         view.setMapType(mapType)
       }
+      prop("zoomControls") { view: GoogleMapsView, enabled: Boolean ->
+        view.setZoomControlButton(enabled)
+      }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]: What do you think about slightly different and in my opinion more clear naming convention ?
```suggestion
      prop("showZoomControls") { view: GoogleMapsView, enabled: Boolean ->
        view.setShowZoomControlButton(enabled)
      }
```

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 2]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
...
+      prop("compass") { view: GoogleMapsView, enabled: Boolean ->
+        view.setCompassButton(enabled)
+      }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]: What do you think about slightly different and in my opinion more clear naming convention ?
```suggestion
      prop("showCompassButton") { view: GoogleMapsView, enabled: Boolean ->
        view.setShowCompassButton(enabled)
      }
```

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 3]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
...
+      prop("mapToolbar") { view: GoogleMapsView, enabled: Boolean ->
+        view.setMapToolbarButton(enabled)
+      }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]:
```suggestion
      prop("showMapToolbarButton") { view: GoogleMapsView, enabled: Boolean ->
        view.setShowMapToolbarButton(enabled)
      }
```

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 4]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
...
+      //TODO: enable MyLocation layer, otherwise My location button will not be displayed
+      prop("myLocationButton") { view: GoogleMapsView, enabled: Boolean ->
+        view.setMyLocationButton(enabled)
+      }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]: Same question as above

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 5]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
...
+      prop("levelPicker") { view: GoogleMapsView, enabled: Boolean ->
+        view.setLevelPickerButton(enabled)
+      }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]: Same question as above

File: expo-maps/build/Maps.types.d.ts [hunk 6]
```diff
@@ -2,11 +2,24 @@ import { ViewProps } from 'react-native';
 import { PropsWithChildren } from 'react';
 export declare type NativeExpoGoogleMapsViewProps = ViewProps & PropsWithChildren<{
     mapType: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    zoomControls: boolean;
+    compass: boolean;
+    mapToolbar: boolean;
+    myLocationButton: boolean;
+    levelPicker: boolean;
```
Comment (szdziedzic) [inline expo-maps/build/Maps.types.d.ts]: Same question as above

File: expo-maps/build/Maps.types.d.ts [hunk 7]
```diff
@@ -2,11 +2,24 @@ import { ViewProps } from 'react-native';
...
 }>;
 export declare type NativeExpoAppleMapsViewProps = ViewProps & PropsWithChildren<{
     mapType: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    compass: boolean;
+    myLocationButton: boolean;
+    levelPicker: boolean;
```
Comment (szdziedzic) [inline expo-maps/build/Maps.types.d.ts]: Same question as above

File: expo-maps/build/Maps.types.d.ts [hunk 8]
```diff
@@ -2,11 +2,24 @@ import { ViewProps } from 'react-native';
...
 }>;
 export declare type ExpoMapViewProps = ViewProps & PropsWithChildren<{
     provider?: 'google' | 'apple';
     mapType?: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    zoomControls?: boolean;
+    compass?: boolean;
+    mapToolbar?: boolean;
+    myLocationButton?: boolean;
+    levelPicker?: boolean;
```
Comment (szdziedzic) [inline expo-maps/build/Maps.types.d.ts]: Same question as above

File: expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift [hunk 9]
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
...
       }
+      prop("compass") { (view: AppleMapsView, enable: Bool) in
+        view.setCompassButton(enable: enable)
+      }
+      prop("myLocationButton") { (view: AppleMapsView, enable: Bool) in
+        view.setMyLocationButton(enable: enable)
+      }
+      prop("levelPicker") { (view: AppleMapsView, enable: Bool) in
+        view.setLevelPickerButton(enable: enable)
+      }
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift]: Same question as above

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 10]
```diff
@@ -45,6 +45,46 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
     }
   }
 
+  fun setZoomControlButton(enabled: Boolean) {
+    updateMap {
+      googleMap.uiSettings.isZoomControlsEnabled = enabled
+    }
+  }
+
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]:
```suggestion
  fun setShowZoomControlButton(enabled: Boolean) {
    updateMap {
      googleMap.uiSettings.isZoomControlsEnabled = enabled
    }
  }

```

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 11]
```diff
@@ -45,6 +45,46 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
...
+      googleMap.uiSettings.isMyLocationButtonEnabled = enabled
+    }
+  }
+
+  fun setLevelPickerButton(enabled: Boolean) {
+    // appears only when the user is viewing an indoor map
+    updateMap {
+      googleMap.uiSettings.isIndoorLevelPickerEnabled = enabled
+    }
+  }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]: Same as above

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 12]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
...
+    
+  func setMyLocationButton(enable: Bool) -> Void {
+    if (enable == true) {
+      self.enableMyLocationButton()
+    }
+  }
+    
+  func setLevelPickerButton(enable: Bool) -> Void {
+    //TODO: enable floor picker
+  }
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]: Same as above

File: expo-maps/ios/ExpoMaps/GoogleMapsView.swift [hunk 13]
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
     }
     self.mapView.mapType = mapViewType
   }
+    
+  func setCompassButton(enabled: Bool) -> Void {
+    self.mapView.settings.compassButton = enabled
+  }
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/GoogleMapsView.swift]:
```suggestion
  func setShowCompassButton(enabled: Bool) -> Void {
    self.mapView.settings.compassButton = enabled
  }
```

File: expo-maps/ios/ExpoMaps/ExpoGoogleMapsModule.swift [hunk 14]
```diff
@@ -11,6 +11,15 @@ public class ExpoGoogleMapsModule: Module {
       prop("mapType") { (view: GoogleMapsView, mapType: MapType) in
         view.setMapType(mapType: mapType)
       }
+      prop ("compass") { (view: GoogleMapsView, enabled: Bool) in
+        view.setCompassButton(enabled: enabled)
+      }
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/ExpoGoogleMapsModule.swift]:
```suggestion
      prop ("showCompassButton") { (view: GoogleMapsView, enabled: Bool) in
        view.setShowCompassButton(enabled: enabled)
      }
```

File: expo-maps/ios/ExpoMaps/ExpoGoogleMapsModule.swift [hunk 15]
```diff
@@ -11,6 +11,15 @@ public class ExpoGoogleMapsModule: Module {
...
+      prop ("myLocationButton") { (view: GoogleMapsView, enabled: Bool) in
+        view.setMyLocationButton(enabled: enabled)
+      }
+      prop ("levelPicker") { (view: GoogleMapsView, enabled: Bool) in
+        view.setFloorPickerButton(enabled: enabled)
+      }
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/ExpoGoogleMapsModule.swift]: Same question as above

File: expo-maps/ios/ExpoMaps/GoogleMapsView.swift [hunk 16]
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
...
+    if (enabled == true) {
+        self.requestLocationPermission()
+        self.mapView.isMyLocationEnabled = true
+    }
+  }
+
+  func setFloorPickerButton(enabled: Bool) -> Void {
+      //appears whenever an indoor map is featured prominently
+    self.mapView.settings.indoorPicker = enabled
+  }
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/GoogleMapsView.swift]: Same question as above

File: expo-maps/src/Maps.tsx [hunk 17]
```diff
@@ -10,10 +10,20 @@ export * from './Maps.types';
 
 interface DefaultNativeExpoMapViewProps {
   mapType: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+  zoomControls: boolean;
+  compass: boolean;
+  mapToolbar: boolean;
+  myLocationButton: boolean;
+  levelPicker: boolean;
```
Comment (szdziedzic) [inline expo-maps/src/Maps.tsx]: Same as above

File: expo-maps/src/Maps.types.ts [hunk 18]
```diff
@@ -4,15 +4,28 @@ import { PropsWithChildren } from 'react';
 export type NativeExpoGoogleMapsViewProps = ViewProps &
   PropsWithChildren<{
     mapType: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    zoomControls: boolean;
+    compass: boolean;
+    mapToolbar: boolean;
+    myLocationButton: boolean;
+    levelPicker: boolean;
```
Comment (szdziedzic) [inline expo-maps/src/Maps.types.ts]: Same as above

File: expo-maps/src/Maps.tsx [hunk 19]
```diff
@@ -10,10 +10,20 @@ export * from './Maps.types';
...
 }
 
 const defaultNativeExpoMapViewProps: DefaultNativeExpoMapViewProps = {
   mapType: 'normal',
+  zoomControls: true,
+  compass: true,
+  mapToolbar: true,
+  myLocationButton: true,
+  levelPicker: true,
```
Comment (szdziedzic) [inline expo-maps/src/Maps.tsx]: Same as above

File: expo-maps/src/Maps.types.ts [hunk 20]
```diff
@@ -4,15 +4,28 @@ import { PropsWithChildren } from 'react';
...
   }>;
 
 export type NativeExpoAppleMapsViewProps = ViewProps &
   PropsWithChildren<{
     mapType: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    compass: boolean;
+    myLocationButton: boolean;
+    levelPicker: boolean;
```
Comment (szdziedzic) [inline expo-maps/src/Maps.types.ts]: Same as above

File: expo-maps/src/Maps.types.ts [hunk 21]
```diff
@@ -4,15 +4,28 @@ import { PropsWithChildren } from 'react';
...
 
 export type ExpoMapViewProps = ViewProps &
   PropsWithChildren<{
     provider?: 'google' | 'apple';
     mapType?: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    zoomControls?: boolean;
+    compass?: boolean;
+    mapToolbar?: boolean;
+    myLocationButton?: boolean;
+    levelPicker?: boolean;
```
Comment (szdziedzic) [inline expo-maps/src/Maps.types.ts]: Same as above

**Review by szdziedzic — COMMENTED** [review]
Just a few suggestions

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 22]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
...
+        myLocationButton.layer.borderWidth = 1
+        myLocationButton.layer.cornerRadius = 5
+        myLocationButton.translatesAutoresizingMaskIntoConstraints = false
+    self.mapView.addSubview(myLocationButton)
+          
+    NSLayoutConstraint.activate([
+        myLocationButton.topAnchor.constraint(equalTo: self.mapView.topAnchor,
+            constant: 100),
+        myLocationButton.trailingAnchor.constraint(equalTo: self.mapView.trailingAnchor, constant: -10)])
+  }
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
    NSLayoutConstraint.activate([
        myLocationButton.topAnchor.constraint(equalTo: self.mapView.topAnchor, constant: 100),
        myLocationButton.trailingAnchor.constraint(equalTo: self.mapView.trailingAnchor, constant: -10)
    ])
  }
```
It looks better in my opinion ;)

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 23]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
...
+    
+  private func requestLocationPermission() -> Void {
+    let locationManager = CLLocationManager()
+    locationManager.requestWhenInUseAuthorization()
+  }
+
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
```

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 24]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
       mapViewType = .standard
     }
     self.mapView.mapType = mapViewType
+
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
```

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 25]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
       mapViewType = .standard
     }
     self.mapView.mapType = mapViewType
+    
   }
+
+  func setCompassButton(enable: Bool) -> Void {
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
  func setCompassButton(enable: Bool) {
```
`Void` type can be omitted.

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 26]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
...
+    self.mapView.showsCompass = enable
+  }
+    
+  func setMyLocationButton(enable: Bool) -> Void {
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
  func setMyLocationButton(enable: Bool) {
```
`Void` type can be omitted.

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 27]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
...
+    if (enable == true) {
+      self.enableMyLocationButton()
+    }
+  }
+    
+  func setLevelPickerButton(enable: Bool) -> Void {
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
  func setLevelPickerButton(enable: Bool) {
```
`Void` type can be omitted.

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 28]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
...
+    //TODO: enable floor picker
+  }
+    
+  private func enableMyLocationButton() -> Void {
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
  private func enableMyLocationButton() {
```
`Void` type can be omitted.

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 29]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
...
+    self.requestLocationPermission()
+
+    self.mapView.showsUserLocation = true
+    let myLocationButton = MKUserTrackingButton(mapView: self.mapView)
+        myLocationButton.layer.backgroundColor = UIColor(white: 1, alpha: 0.5).cgColor
+        myLocationButton.layer.borderColor = UIColor.white.cgColor
+        myLocationButton.layer.borderWidth = 1
+        myLocationButton.layer.cornerRadius = 5
+        myLocationButton.translatesAutoresizingMaskIntoConstraints = false
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
    let myLocationButton = MKUserTrackingButton(mapView: self.mapView)
    myLocationButton.layer.backgroundColor = UIColor(white: 1, alpha: 0.5).cgColor
    myLocationButton.layer.borderColor = UIColor.white.cgColor
    myLocationButton.layer.borderWidth = 1
    myLocationButton.layer.cornerRadius = 5
    myLocationButton.translatesAutoresizingMaskIntoConstraints = false
```

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 30]
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
...
+    self.mapView.addSubview(myLocationButton)
+          
+    NSLayoutConstraint.activate([
+        myLocationButton.topAnchor.constraint(equalTo: self.mapView.topAnchor,
+            constant: 100),
+        myLocationButton.trailingAnchor.constraint(equalTo: self.mapView.trailingAnchor, constant: -10)])
+  }
+    
+  private func requestLocationPermission() -> Void {
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]:
```suggestion
  private func requestLocationPermission() {
```

File: expo-maps/ios/ExpoMaps/GoogleMapsView.swift [hunk 31]
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
...
+  func setMyLocationButton(enabled: Bool) -> Void {
+    self.mapView.settings.myLocationButton = enabled
+    if (enabled == true) {
+        self.requestLocationPermission()
+        self.mapView.isMyLocationEnabled = true
+    }
+  }
+
+  func setFloorPickerButton(enabled: Bool) -> Void {
+      //appears whenever an indoor map is featured prominently
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/GoogleMapsView.swift]:
```suggestion
    //appears whenever an indoor map is featured prominently
```

File: expo-maps/ios/ExpoMaps/GoogleMapsView.swift [hunk 32]
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
...
+    
+  func setCompassButton(enabled: Bool) -> Void {
+    self.mapView.settings.compassButton = enabled
+  }
+
+  func setMyLocationButton(enabled: Bool) -> Void {
+    self.mapView.settings.myLocationButton = enabled
+    if (enabled == true) {
+        self.requestLocationPermission()
+        self.mapView.isMyLocationEnabled = true
```
Comment (lukmccall) [inline expo-maps/ios/ExpoMaps/GoogleMapsView.swift]:
```suggestion
      self.requestLocationPermission()
      self.mapView.isMyLocationEnabled = true
```

**Review by lukmccall — COMMENTED** [review]
Looks good to me 🏅 
Just a couple of comments: 
- I would choose the naming proposed by Szymon.
- You don't have to ask for location permission manually. Please install `expo-location` (https://github.com/expo/expo/tree/master/packages/expo-location#installation-in-bare-react-native-projects) and call `Location.requestForegroundPermissions` via js. We can add a native permission requester later.

File: expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt [hunk 33]
```diff
@@ -45,6 +45,46 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
...
+
+  fun setShowLevelPickerButton(enabled: Boolean) {
+    // appears only when the user is viewing an indoor map
+    updateMap {
+      googleMap.uiSettings.isIndoorLevelPickerEnabled = enabled
+    }
+  }
+
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt]: I'm wondering if it wouldn't be more consistent solution to move this logic into `GoogleMapsControls `class as I did in my [markers PR](https://github.com/software-mansion-labs/expo-maps/pull/17) or @Zwatotem did in his PR for gestures. What's your opinion?

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 34]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
       prop("mapType") { view: GoogleMapsView, mapType: MapType ->
         view.setMapType(mapType)
       }
+      prop("showZoomControls") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowZoomControlButton(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]:
```suggestion
        view.setShowZoomControls(enabled)
```

Maybe we should try to rename functions to match prop names. What do you think? Overall everything looks good here, it's just cosmetic change suggestion.

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 35]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
...
+      }
+      prop("showCompass") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowCompassButton(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]: As above

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 36]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
...
+      }
+      prop("showMapToolbar") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowMapToolbarButton(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]: As above

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 37]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
...
+      }
+      //TODO: enable MyLocation layer, otherwise My location button will not be displayed
+      prop("showMyLocationButton") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowMyLocationButton(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]: As above

File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt [hunk 38]
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
...
+      }
+      prop("showLevelPicker") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowLevelPickerButton(enabled)
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt]: As above

File: expo-maps/ios/ExpoMaps/AppleMapsView.swift [hunk 39]
```diff
@@ -27,4 +27,43 @@ public final class AppleMapsView: UIView {
...
+constant: 100),
+        myLocationButton.trailingAnchor.constraint(equalTo:
+self.mapView.trailingAnchor, constant: -10)
+    ])
+  }
+    
+  private func requestLocationPermission() {
+    let locationManager = CLLocationManager()
+    locationManager.requestWhenInUseAuthorization()
+  }
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/AppleMapsView.swift]: As above, we can try to move this logic to `AppleMapsControls` class

File: expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift [hunk 40]
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
       prop("mapType") { (view: AppleMapsView, mapType: MapType) in
         view.setMapType(mapType: mapType)
       }
+      prop("showCmpass") { (view: AppleMapsView, enable: Bool) in
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift]:
```suggestion
      prop("showCompass") { (view: AppleMapsView, enable: Bool) in
```

File: expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift [hunk 41]
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
...
+        view.setShowCompassButton(enable: enable)
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift]: Same suggestion as for Android
```suggestion
        view.setShowCompass(enable: enable)
```

File: expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift [hunk 42]
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
...
+      }
+      prop("showMyLocationButton") { (view: AppleMapsView, enable: Bool) in
+        view.setShowMyLocationButton(enable: enable)
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift:50]: As above

File: expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift [hunk 43]
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
...
+      }
+      prop("showLevelPicker") { (view: AppleMapsView, enable: Bool) in
+        view.setShowLevelPickerButton(enable: enable)
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift]: As above

File: expo-maps/ios/ExpoMaps/GoogleMapsView.swift [hunk 44]
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
...
+
+  func setShowFloorPickerButton(enabled: Bool) {
+    //appears whenever an indoor map is featured prominently
+    self.mapView.settings.indoorPicker = enabled
+  }
+    
+    private func requestLocationPermission() {
+      let locationManager = CLLocationManager()
+      locationManager.requestWhenInUseAuthorization()
+    }
```
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/GoogleMapsView.swift]: We can also try to move this logic to `GoogleMapsControls` class

**Review by szdziedzic — COMMENTED** [review]
Good job 🚀! Everything looks correct 🎉! I'm so sorry about being so nitpicky, but I left few more cosmetic suggestions 😅

**Comment by ellwiii** [comment]
Update from me:
I have applied changes proposed by @szdziedzic:

- changed functions naming to mach pops names
- unified "enable" parameters name
- moved logic of setting controllers into new classes

About expo-location setup - I've not yet figured out the correct way to use it. I was trying to prompt user for allowing location permissions via calling requestForegroundPermissionsAsync() in App.js but it didn't provide the expected result. I will try to resolve this issue in the following days.

**Comment by ellwiii** [comment]
I enabled getting location permissions from users via expo-location (the previous problem turned out to be you have to manually reset location permissions settings on iOS simulator and Android emulator - that's why the "ask for permissions" window was not visible to me). As @lukmccall recommended, this solution also in my opinion is fine for now ☺️

**Review by Zwatotem — COMMENTED** [review]
I don't see any protocol added to bond GoogleMapsControllers and AppleMapsControlles with a common interface, like it is done on the Android side. This seems counter-intuitive to have an interface, where there is one implementation and no interface, when there's multiple.

**Comment by ellwiii** [comment]
> I don't see any protocol added to bond GoogleMapsControllers and AppleMapsControlles with a common interface, like it is done on the Android side. This seems counter-intuitive to have an interface, where there is one implementation and no interface, when there's multiple.

You are completely right @Zwatotem, I have omitted this in my implementation. Now it's added.

File: expo-maps/android/src/main/java/expo/modules/maps/Controllers.kt [hunk 45]
```diff
@@ -0,0 +1,9 @@
+package expo.modules.maps
+
+interface Controllers {
+    fun setShowZoomControl(enable: Boolean)
+    fun setShowCompass(enable: Boolean)
+    fun setShowMapToolbar(enable: Boolean)
+    fun setShowMyLocationButton(enable: Boolean)
+    fun setShowLevelPicker(enable: Boolean)
+}
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/Controllers.kt]:
```suggestion
}

```

File: example/App.tsx [hunk 46]
```diff
@@ -1,10 +1,22 @@
...
 
 export default function App() {
+
+  const getLocationPermissions = async () => {
+    await Location.requestForegroundPermissionsAsync();
+  }
+
+  useEffect(() => {
+    getLocationPermissions();
+  }, []);
```
Comment (szdziedzic) [inline example/App.tsx:18]: I'm wondering if user should do this in his App before using `Expo Maps ` or maybe if we should do this as a part of library functionalities. What do you think?

File: expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift (hunk 40 above)
Comment (szdziedzic) [inline expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift]:
```suggestion
      prop("showCompass") { (view: AppleMapsView, enable: Bool) in
```

I think this is typo

**Review by szdziedzic — APPROVED** [review]
Nice one. Looks alright 👍. It's good to go in my opinion 👏 . I leave few last comments for you:

File: example/App.tsx (hunk 46 above)
Comment (ellwiii) [reply example/App.tsx:18]: @szdziedzic This is only a temporary solution. I believe finally it should be a part of library functionalities.

In my opinion we should create separate issue to work on this topic. The solution will also depend strongly whether we'd like to keep using expo-location or switch to native ways of getting location permissions from user.

Comment (szdziedzic) [reply example/App.tsx:18]: Sounds reasonable 👍

**Review by lukmccall — APPROVED** [review]

**Pull request merged**